# Data Preparation Phase

In this notebook, we will transform our raw data into a machine-learning-ready format. The steps include:
1. Loading and examining the dataset.
2. Removing exact duplicates to ensure data integrity.
3. Splitting the data into Training and Testing sets *before* any transformations to prevent data leakage.
4. Applying feature scaling to normalize the input variables.

*Note: We are intentionally bypassing Principal Component Analysis (PCA) to maintain 100% feature interpretability. Our tree-based models can efficiently handle the original 8-dimensional feature space, allowing us to eventually extract direct feature importances.*

In [24]:
# Import required libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import StandardScaler

import os
import shutil

# Set display options
pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', lambda x: '%.3f' % x)

# Set plotting style - using a simpler style to avoid seaborn issues
plt.style.use('default')

# Ensure directories exist
os.makedirs('../data/interim', exist_ok=True)
os.makedirs('../data/processed', exist_ok=True)

# Copy raw data to interim folder if it doesn't exist there
if not os.path.exists('../data/interim/concrete_data.csv'):
    shutil.copy('../data/raw/Concrete_Data_Yeh.csv', '../data/interim/concrete_data.csv')

# Load the dataset
print("Loading dataset...")
df = pd.read_csv('../data/interim/concrete_data.csv')

print("Dataset loaded successfully!")
print(f"Shape: {df.shape}")
print("\nColumns in the dataset:")
print(df.columns.tolist())
print("\nFirst few rows:")
print(df.head())


Loading dataset...
Dataset loaded successfully!
Shape: (1030, 9)

Columns in the dataset:
['cement', 'slag', 'flyash', 'water', 'superplasticizer', 'coarseaggregate', 'fineaggregate', 'age', 'csMPa']

First few rows:
   cement    slag  flyash   water  superplasticizer  coarseaggregate  \
0 540.000   0.000   0.000 162.000             2.500         1040.000   
1 540.000   0.000   0.000 162.000             2.500         1055.000   
2 332.500 142.500   0.000 228.000             0.000          932.000   
3 332.500 142.500   0.000 228.000             0.000          932.000   
4 198.600 132.400   0.000 192.000             0.000          978.400   

   fineaggregate  age  csMPa  
0        676.000   28 79.990  
1        676.000   28 61.890  
2        594.000  270 40.270  
3        594.000  365 41.050  
4        825.500  360 44.300  


In [25]:
# Display detailed information about the columns
print("Column names in the dataset:")
for col in df.columns:
    print(f"'{col}'")
    
print("\nDataset Info:")
df.info()


Column names in the dataset:
'cement'
'slag'
'flyash'
'water'
'superplasticizer'
'coarseaggregate'
'fineaggregate'
'age'
'csMPa'

Dataset Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1030 entries, 0 to 1029
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   cement            1030 non-null   float64
 1   slag              1030 non-null   float64
 2   flyash            1030 non-null   float64
 3   water             1030 non-null   float64
 4   superplasticizer  1030 non-null   float64
 5   coarseaggregate   1030 non-null   float64
 6   fineaggregate     1030 non-null   float64
 7   age               1030 non-null   int64  
 8   csMPa             1030 non-null   float64
dtypes: float64(8), int64(1)
memory usage: 72.5 KB


## 2. Handle Exact Duplicates

Let's check for and remove exact duplicates in the dataset.


In [26]:
# Check for duplicates
initial_rows = len(df)
duplicates = df.duplicated()
n_duplicates = duplicates.sum()

print(f"Number of exact duplicates found: {n_duplicates}")

if n_duplicates > 0:
    # Remove duplicates
    df = df.drop_duplicates()
    print(f"\nRows after removing duplicates: {len(df)}")
    print(f"Removed {initial_rows - len(df)} duplicate rows")

# Save intermediate result
df.to_csv('../data/interim/concrete_data_no_duplicates.csv', index=False)
print("\nIntermediate dataset saved to '../data/interim/concrete_data_no_duplicates.csv'")


Number of exact duplicates found: 25

Rows after removing duplicates: 1005
Removed 25 duplicate rows

Intermediate dataset saved to '../data/interim/concrete_data_no_duplicates.csv'


## 3. Feature Scaling

We'll use StandardScaler for feature scaling. We'll scale all features except the target variable.


In [27]:
# Find the target column (looking for 'MPa' or 'strength' in column name)
target_col = "csMPa"  # Default target column name

# we divide train and test sets.
from sklearn.model_selection import train_test_split
train_set, test_set = train_test_split(df, test_size=0.2, random_state=42)

# we save the train and test sets for later use
train_set.to_csv('../data/interim/concrete_train_set.csv', index=False)
test_set.to_csv('../data/interim/concrete_test_set.csv', index=False)

# Get feature columns (all columns except target)
X_train = train_set.drop(columns=[target_col])
y_train = train_set[target_col]

X_test = test_set.drop(columns=[target_col])
y_test = test_set[target_col]

# Initialize scaler
scaler = StandardScaler()
# we save the scaler for later use on test set and new data



# Scale features
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)
import joblib
joblib.dump(scaler, '../models/standard_scaler.pkl')
# Print scaling report
print("\nFeature Scaling Report:")
print("-" * 50)
#for column in X_train.columns:
    #t(f"\nFeature: {column}")
    #t("Before scaling:")
    #t(f"  Mean: {X_train[column].mean():.3f}")
    #t(f"  Std: {X_train[column].std():.3f}")
    #print("After scaling:")
    #print(f"  Mean: {X_train_scaled[column].mean():.2f}")
    #print(f"  Std: {X_train_scaled[column].std():.3f}")

X_train_scaled_df = pd.DataFrame(X_train_scaled, columns=X_train.columns)
print("\nScaled features (first 5 rows):")
print(X_train_scaled_df.head())
X_train_scaled_df.to_csv('../data/interim/concrete_data_scaled.csv', index=False)
print("\nScaled dataset saved to '../data/interim/concrete_data_scaled.csv'")

X_test_scaled_df = pd.DataFrame(X_test_scaled, columns=X_test.columns)
X_test_scaled_df.to_csv('../data/interim/concrete_test_data_scaled.csv', index=False)
print("\nScaled test dataset saved to '../data/interim/concrete_test_data_scaled.csv'")

#now we savew everything for later use in modelling and evaluation
train_set_scaled = X_train_scaled_df.copy()
train_set_scaled[target_col] = y_train.values
train_set_scaled.to_csv('../data/processed/concrete_train_set_scaled.csv', index=False)
print("\nFinal processed training set saved to '../data/processed/concrete_train_set_scaled.csv'")
test_set_scaled = X_test_scaled_df.copy()
test_set_scaled[target_col] = y_test.values
test_set_scaled.to_csv('../data/processed/concrete_test_set_scaled.csv', index=False)
print("\nFinal processed test set saved to '../data/processed/concrete_test_set_scaled.csv'")


Feature Scaling Report:
--------------------------------------------------

Scaled features (first 5 rows):
   cement   slag  flyash  water  superplasticizer  coarseaggregate  \
0   2.437 -0.838  -0.876 -1.873             3.720           -1.560   
1   1.895 -0.838  -0.876  2.132            -1.023           -0.526   
2  -0.010 -0.838   0.639 -0.999             0.962            0.017   
3   1.519 -0.838  -0.876  1.668            -1.023           -1.737   
4  -1.072  2.070  -0.876  0.994            -1.023            0.039   

   fineaggregate    age  
0          1.487 -0.679  
1         -2.219 -0.615  
2          1.263 -0.679  
3         -0.664 -0.278  
4         -0.999 -0.615  

Scaled dataset saved to '../data/interim/concrete_data_scaled.csv'

Scaled test dataset saved to '../data/interim/concrete_test_data_scaled.csv'

Final processed training set saved to '../data/processed/concrete_train_set_scaled.csv'

Final processed test set saved to '../data/processed/concrete_test_set_scaled.

## Data Preparation Summary

* **Duplicate Removal:** We successfully dropped the 25 exact duplicates, reducing the dataset from 1030 to 1005 clean rows.
* **Data Splitting:** The data was split into an 80/20 train-test split *prior* to scaling. This is a critical step to ensure the model has no prior knowledge of the test set's distribution.
* **Feature Scaling:** We applied a `StandardScaler` to the 8 input features. The scaler was fitted exclusively on the training data. The target variable (`csMPa`) was left unscaled to keep predictions in their native Megapascal units.
* **Pipeline Consistency:** While tree-based models do not mathematically require scaling, Linear Regression does. We scaled the dataset to ensure a fair, apples-to-apples comparison across all algorithms.